File: standalone/lattice-core/haskell/Lattice/Schemas/AI/MotionSuggestionSchema.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : Lattice.Schemas.AI.MotionSuggestionSchema
Description : AI motion suggestion validation
Copyright   : (c) Lattice, 2026
License     : MIT

Validates AI motion suggestion responses for camera/layer/particle/spline intents.

Source: ui/src/schemas/ai/motion-suggestion-schema.ts
-}

module Lattice.Schemas.AI.MotionSuggestionSchema
  ( -- * Camera Motion Type
    CameraMotionType(..)
  , cameraMotionTypeFromText
  , cameraMotionTypeToText
    -- * Motion Intensity
  , MotionIntensity(..)
  , motionIntensityFromText
  , motionIntensityToText
    -- * Easing Type
  , EasingType(..)
  , easingTypeFromText
  , easingTypeToText
    -- * Spline Usage
  , SplineUsage(..)
  , splineUsageFromText
  , splineUsageToText
    -- * Particle Behavior
  , ParticleBehavior(..)
  , particleBehaviorFromText
  , particleBehaviorToText
    -- * Layer Motion Type
  , LayerMotionType(..)
  , layerMotionTypeFromText
  , layerMotionTypeToText
    -- * Color Scheme
  , ColorScheme(..)
  , colorSchemeFromText
  , colorSchemeToText
    -- * Control Point Type
  , ControlPointType(..)
  , controlPointTypeFromText
  , controlPointTypeToText
    -- * Axis
  , Axis(..)
  , axisFromText
  , axisToText
    -- * Motion Axis
  , MotionAxis(..)
  , motionAxisFromText
  , motionAxisToText
  ) where

import GHC.Generics (Generic)
import Data.Text (Text)

-- ────────────────────────────────────────────────────────────────────────────
-- Camera Motion Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Camera motion type options
data CameraMotionType
  = CamDolly
  | CamTruck
  | CamPedestal
  | CamPan
  | CamTilt
  | CamRoll
  | CamOrbit
  | CamDrift
  | CamHandheld
  | CamCrane
  | CamZoom
  | CamFollowPath
  deriving stock (Eq, Show, Generic, Enum, Bounded)

cameraMotionTypeFromText :: Text -> Maybe CameraMotionType
cameraMotionTypeFromText "dolly" = Just CamDolly
cameraMotionTypeFromText "truck" = Just CamTruck
cameraMotionTypeFromText "pedestal" = Just CamPedestal
cameraMotionTypeFromText "pan" = Just CamPan
cameraMotionTypeFromText "tilt" = Just CamTilt
cameraMotionTypeFromText "roll" = Just CamRoll
cameraMotionTypeFromText "orbit" = Just CamOrbit
cameraMotionTypeFromText "drift" = Just CamDrift
cameraMotionTypeFromText "handheld" = Just CamHandheld
cameraMotionTypeFromText "crane" = Just CamCrane
cameraMotionTypeFromText "zoom" = Just CamZoom
cameraMotionTypeFromText "follow_path" = Just CamFollowPath
cameraMotionTypeFromText _ = Nothing

cameraMotionTypeToText :: CameraMotionType -> Text
cameraMotionTypeToText CamDolly = "dolly"
cameraMotionTypeToText CamTruck = "truck"
cameraMotionTypeToText CamPedestal = "pedestal"
cameraMotionTypeToText CamPan = "pan"
cameraMotionTypeToText CamTilt = "tilt"
cameraMotionTypeToText CamRoll = "roll"
cameraMotionTypeToText CamOrbit = "orbit"
cameraMotionTypeToText CamDrift = "drift"
cameraMotionTypeToText CamHandheld = "handheld"
cameraMotionTypeToText CamCrane = "crane"
cameraMotionTypeToText CamZoom = "zoom"
cameraMotionTypeToText CamFollowPath = "follow_path"

-- ────────────────────────────────────────────────────────────────────────────
-- Motion Intensity
-- ────────────────────────────────────────────────────────────────────────────

-- | Motion intensity levels
data MotionIntensity
  = IntensityVerySubtle
  | IntensitySubtle
  | IntensityMedium
  | IntensityStrong
  | IntensityDramatic
  deriving stock (Eq, Show, Generic, Enum, Bounded)

motionIntensityFromText :: Text -> Maybe MotionIntensity
motionIntensityFromText "very_subtle" = Just IntensityVerySubtle
motionIntensityFromText "subtle" = Just IntensitySubtle
motionIntensityFromText "medium" = Just IntensityMedium
motionIntensityFromText "strong" = Just IntensityStrong
motionIntensityFromText "dramatic" = Just IntensityDramatic
motionIntensityFromText _ = Nothing

motionIntensityToText :: MotionIntensity -> Text
motionIntensityToText IntensityVerySubtle = "very_subtle"
motionIntensityToText IntensitySubtle = "subtle"
motionIntensityToText IntensityMedium = "medium"
motionIntensityToText IntensityStrong = "strong"
motionIntensityToText IntensityDramatic = "dramatic"

-- ────────────────────────────────────────────────────────────────────────────
-- Easing Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Easing type options
data EasingType
  = EaseLinear
  | EaseIn
  | EaseOut
  | EaseInOut
  | EaseBounce
  | EaseElastic
  deriving stock (Eq, Show, Generic, Enum, Bounded)

easingTypeFromText :: Text -> Maybe EasingType
easingTypeFromText "linear" = Just EaseLinear
easingTypeFromText "easeIn" = Just EaseIn
easingTypeFromText "easeOut" = Just EaseOut
easingTypeFromText "easeInOut" = Just EaseInOut
easingTypeFromText "bounce" = Just EaseBounce
easingTypeFromText "elastic" = Just EaseElastic
easingTypeFromText _ = Nothing

easingTypeToText :: EasingType -> Text
easingTypeToText EaseLinear = "linear"
easingTypeToText EaseIn = "easeIn"
easingTypeToText EaseOut = "easeOut"
easingTypeToText EaseInOut = "easeInOut"
easingTypeToText EaseBounce = "bounce"
easingTypeToText EaseElastic = "elastic"

-- ────────────────────────────────────────────────────────────────────────────
-- Spline Usage
-- ────────────────────────────────────────────────────────────────────────────

-- | Spline usage options
data SplineUsage
  = SplineCameraPath
  | SplineEmitterPath
  | SplineTextPath
  | SplineLayerPath
  deriving stock (Eq, Show, Generic, Enum, Bounded)

splineUsageFromText :: Text -> Maybe SplineUsage
splineUsageFromText "camera_path" = Just SplineCameraPath
splineUsageFromText "emitter_path" = Just SplineEmitterPath
splineUsageFromText "text_path" = Just SplineTextPath
splineUsageFromText "layer_path" = Just SplineLayerPath
splineUsageFromText _ = Nothing

splineUsageToText :: SplineUsage -> Text
splineUsageToText SplineCameraPath = "camera_path"
splineUsageToText SplineEmitterPath = "emitter_path"
splineUsageToText SplineTextPath = "text_path"
splineUsageToText SplineLayerPath = "layer_path"

-- ────────────────────────────────────────────────────────────────────────────
-- Particle Behavior
-- ────────────────────────────────────────────────────────────────────────────

-- | Particle behavior options
data ParticleBehavior
  = BehaviorFlow
  | BehaviorDrift
  | BehaviorSpray
  | BehaviorTurbulence
  | BehaviorExplosion
  | BehaviorVortex
  | BehaviorRain
  | BehaviorSnow
  | BehaviorFireflies
  | BehaviorDust
  | BehaviorAlongPath
  deriving stock (Eq, Show, Generic, Enum, Bounded)

particleBehaviorFromText :: Text -> Maybe ParticleBehavior
particleBehaviorFromText "flow" = Just BehaviorFlow
particleBehaviorFromText "drift" = Just BehaviorDrift
particleBehaviorFromText "spray" = Just BehaviorSpray
particleBehaviorFromText "turbulence" = Just BehaviorTurbulence
particleBehaviorFromText "explosion" = Just BehaviorExplosion
particleBehaviorFromText "vortex" = Just BehaviorVortex
particleBehaviorFromText "rain" = Just BehaviorRain
particleBehaviorFromText "snow" = Just BehaviorSnow
particleBehaviorFromText "fireflies" = Just BehaviorFireflies
particleBehaviorFromText "dust" = Just BehaviorDust
particleBehaviorFromText "along_path" = Just BehaviorAlongPath
particleBehaviorFromText _ = Nothing

particleBehaviorToText :: ParticleBehavior -> Text
particleBehaviorToText BehaviorFlow = "flow"
particleBehaviorToText BehaviorDrift = "drift"
particleBehaviorToText BehaviorSpray = "spray"
particleBehaviorToText BehaviorTurbulence = "turbulence"
particleBehaviorToText BehaviorExplosion = "explosion"
particleBehaviorToText BehaviorVortex = "vortex"
particleBehaviorToText BehaviorRain = "rain"
particleBehaviorToText BehaviorSnow = "snow"
particleBehaviorToText BehaviorFireflies = "fireflies"
particleBehaviorToText BehaviorDust = "dust"
particleBehaviorToText BehaviorAlongPath = "along_path"

-- ────────────────────────────────────────────────────────────────────────────
-- Layer Motion Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Layer motion type options
data LayerMotionType
  = LayerParallax
  | LayerFloat
  | LayerSway
  | LayerBreathe
  | LayerDrift
  | LayerNoise
  | LayerPulse
  | LayerRotate
  | LayerFollowPath
  deriving stock (Eq, Show, Generic, Enum, Bounded)

layerMotionTypeFromText :: Text -> Maybe LayerMotionType
layerMotionTypeFromText "parallax" = Just LayerParallax
layerMotionTypeFromText "float" = Just LayerFloat
layerMotionTypeFromText "sway" = Just LayerSway
layerMotionTypeFromText "breathe" = Just LayerBreathe
layerMotionTypeFromText "drift" = Just LayerDrift
layerMotionTypeFromText "noise" = Just LayerNoise
layerMotionTypeFromText "pulse" = Just LayerPulse
layerMotionTypeFromText "rotate" = Just LayerRotate
layerMotionTypeFromText "follow_path" = Just LayerFollowPath
layerMotionTypeFromText _ = Nothing

layerMotionTypeToText :: LayerMotionType -> Text
layerMotionTypeToText LayerParallax = "parallax"
layerMotionTypeToText LayerFloat = "float"
layerMotionTypeToText LayerSway = "sway"
layerMotionTypeToText LayerBreathe = "breathe"
layerMotionTypeToText LayerDrift = "drift"
layerMotionTypeToText LayerNoise = "noise"
layerMotionTypeToText LayerPulse = "pulse"
layerMotionTypeToText LayerRotate = "rotate"
layerMotionTypeToText LayerFollowPath = "follow_path"

-- ────────────────────────────────────────────────────────────────────────────
-- Color Scheme
-- ────────────────────────────────────────────────────────────────────────────

-- | Color scheme options
data ColorScheme = SchemeWarm | SchemeCool | SchemeNeutral | SchemeCustom
  deriving stock (Eq, Show, Generic, Enum, Bounded)

colorSchemeFromText :: Text -> Maybe ColorScheme
colorSchemeFromText "warm" = Just SchemeWarm
colorSchemeFromText "cool" = Just SchemeCool
colorSchemeFromText "neutral" = Just SchemeNeutral
colorSchemeFromText "custom" = Just SchemeCustom
colorSchemeFromText _ = Nothing

colorSchemeToText :: ColorScheme -> Text
colorSchemeToText SchemeWarm = "warm"
colorSchemeToText SchemeCool = "cool"
colorSchemeToText SchemeNeutral = "neutral"
colorSchemeToText SchemeCustom = "custom"

-- ────────────────────────────────────────────────────────────────────────────
-- Control Point Type
-- ────────────────────────────────────────────────────────────────────────────

-- | Control point type options
data ControlPointType = PointCorner | PointSmooth | PointSymmetric
  deriving stock (Eq, Show, Generic, Enum, Bounded)

controlPointTypeFromText :: Text -> Maybe ControlPointType
controlPointTypeFromText "corner" = Just PointCorner
controlPointTypeFromText "smooth" = Just PointSmooth
controlPointTypeFromText "symmetric" = Just PointSymmetric
controlPointTypeFromText _ = Nothing

controlPointTypeToText :: ControlPointType -> Text
controlPointTypeToText PointCorner = "corner"
controlPointTypeToText PointSmooth = "smooth"
controlPointTypeToText PointSymmetric = "symmetric"

-- ────────────────────────────────────────────────────────────────────────────
-- Axis
-- ────────────────────────────────────────────────────────────────────────────

-- | Axis options
data Axis = AxisX | AxisY | AxisZ | AxisAll
  deriving stock (Eq, Show, Generic, Enum, Bounded)

axisFromText :: Text -> Maybe Axis
axisFromText "x" = Just AxisX
axisFromText "y" = Just AxisY
axisFromText "z" = Just AxisZ
axisFromText "all" = Just AxisAll
axisFromText _ = Nothing

axisToText :: Axis -> Text
axisToText AxisX = "x"
axisToText AxisY = "y"
axisToText AxisZ = "z"
axisToText AxisAll = "all"

-- ────────────────────────────────────────────────────────────────────────────
-- Motion Axis
-- ────────────────────────────────────────────────────────────────────────────

-- | Motion axis options
data MotionAxis = MAxisX | MAxisY | MAxisZ | MAxisScale | MAxisRotation
  deriving stock (Eq, Show, Generic, Enum, Bounded)

motionAxisFromText :: Text -> Maybe MotionAxis
motionAxisFromText "x" = Just MAxisX
motionAxisFromText "y" = Just MAxisY
motionAxisFromText "z" = Just MAxisZ
motionAxisFromText "scale" = Just MAxisScale
motionAxisFromText "rotation" = Just MAxisRotation
motionAxisFromText _ = Nothing

motionAxisToText :: MotionAxis -> Text
motionAxisToText MAxisX = "x"
motionAxisToText MAxisY = "y"
motionAxisToText MAxisZ = "z"
motionAxisToText MAxisScale = "scale"
motionAxisToText MAxisRotation = "rotation"

File: comfyui-extension/ui/src/schemas/ai/motion-suggestion-schema.ts
/**
 * Motion Suggestion Schema
 *
 * Zod schemas for validating AI motion suggestion responses.
 * Used in visionAuthoring/MotionIntentResolver for parsing AI responses.
 * Includes comprehensive validation constraints for security and data integrity.
 */

import { z } from "zod";
import {
  entityId,
  boundedArray,
  MAX_DESCRIPTION_LENGTH,
} from "../shared-validation";
import { Vec3Schema } from "../layers/primitives-schema";
import { normalized01, finiteNumber, positiveFinite, nonNegativeFinite } from "../layers/primitives-schema";

// ============================================================================
// Primitives
// ============================================================================

const positiveInt = z.number().int().positive();

// ============================================================================
// Control Point Schema
// ============================================================================

/**
 * Bezier handle (2D, not 3D)
 */
const BezierHandleSchema = z.object({
  x: finiteNumber,
  y: finiteNumber,
}).strict();

/**
 * Control point for spline/path suggestions
 * Matches ControlPoint from types/spline.ts
 */
export const ControlPointSchema = z.object({
  id: z.string().min(1).max(200).trim(),
  x: finiteNumber,
  y: finiteNumber,
  depth: normalized01.optional(), // Normalized depth 0-1
  handleIn: BezierHandleSchema.nullable().optional(),
  handleOut: BezierHandleSchema.nullable().optional(),
  type: z.enum(["corner", "smooth", "symmetric"]).optional(),
  group: z.string().max(200).trim().optional(), // Group ID for grouped animations
}).strict();

export type ControlPoint = z.infer<typeof ControlPointSchema>;

// ============================================================================
// Camera Motion Type Schema
// ============================================================================

export const CameraMotionTypeSchema = z.enum([
  "dolly",
  "truck",
  "pedestal",
  "pan",
  "tilt",
  "roll",
  "orbit",
  "drift",
  "handheld",
  "crane",
  "zoom",
  "follow_path",
]);

export type CameraMotionType = z.infer<typeof CameraMotionTypeSchema>;

// ============================================================================
// Motion Intensity Schema
// ============================================================================

export const MotionIntensitySchema = z.enum([
  "very_subtle",
  "subtle",
  "medium",
  "strong",
  "dramatic",
]);

export type MotionIntensity = z.infer<typeof MotionIntensitySchema>;

// ============================================================================
// Easing Type Schema
// ============================================================================

export const EasingTypeSchema = z.enum([
  "linear",
  "easeIn",
  "easeOut",
  "easeInOut",
  "bounce",
  "elastic",
]);

export type EasingType = z.infer<typeof EasingTypeSchema>;

// ============================================================================
// Camera Motion Intent Schema
// ============================================================================

export const CameraMotionIntentSchema = z.object({
  type: CameraMotionTypeSchema,
  intensity: MotionIntensitySchema,
  axis: z.enum(["x", "y", "z", "all"]).optional(),
  durationFrames: positiveInt.max(100000).optional(), // Max 100k frames
  suggestedEasing: EasingTypeSchema.optional(),
  noiseAmount: normalized01.optional(), // For handheld
  orbitCenter: Vec3Schema.optional(), // For orbit
  suggestedPath: boundedArray(ControlPointSchema, 10000).optional(), // Max 10k points
}).strict();

export type CameraMotionIntent = z.infer<typeof CameraMotionIntentSchema>;

// ============================================================================
// Spline Usage Schema
// ============================================================================

export const SplineUsageSchema = z.enum([
  "camera_path",
  "emitter_path",
  "text_path",
  "layer_path",
]);

export type SplineUsage = z.infer<typeof SplineUsageSchema>;

// ============================================================================
// Spline Motion Intent Schema
// ============================================================================

export const SplineMotionIntentSchema = z.object({
  usage: SplineUsageSchema,
  smoothness: normalized01, // 0-1 curve smoothness
  complexity: positiveInt.max(10000).optional(), // Max 10k control points
  worldSpace: z.boolean(),
  suggestedPoints: boundedArray(ControlPointSchema, 10000), // Max 10k points
  closed: z.boolean(),
}).strict();

export type SplineMotionIntent = z.infer<typeof SplineMotionIntentSchema>;

// ============================================================================
// Particle Behavior Schema
// ============================================================================

export const ParticleBehaviorSchema = z.enum([
  "flow",
  "drift",
  "spray",
  "turbulence",
  "explosion",
  "vortex",
  "rain",
  "snow",
  "fireflies",
  "dust",
  "along_path",
]);

export type ParticleBehavior = z.infer<typeof ParticleBehaviorSchema>;

// ============================================================================
// Particle Motion Intent Schema
// ============================================================================

export const ParticleMotionIntentSchema = z.object({
  behavior: ParticleBehaviorSchema,
  direction: Vec3Schema.optional(),
  intensity: normalized01, // 0-1 intensity
  spread: normalized01.optional(), // 0-1 spread
  lifetime: positiveFinite.max(3600).optional(), // Max 1 hour lifetime
  colorScheme: z.enum(["warm", "cool", "neutral", "custom"]).optional(),
  suggestedEmitPath: boundedArray(ControlPointSchema, 10000).optional(), // Max 10k points
}).strict();

export type ParticleMotionIntent = z.infer<typeof ParticleMotionIntentSchema>;

// ============================================================================
// Layer Motion Type Schema
// ============================================================================

export const LayerMotionTypeSchema = z.enum([
  "parallax",
  "float",
  "sway",
  "breathe",
  "drift",
  "noise",
  "pulse",
  "rotate",
  "follow_path",
]);

export type LayerMotionType = z.infer<typeof LayerMotionTypeSchema>;

// ============================================================================
// Layer Motion Intent Schema
// ============================================================================

export const LayerMotionIntentSchema = z.object({
  targetLayerId: entityId.optional(),
  motionType: LayerMotionTypeSchema,
  amplitude: normalized01, // 0-1 amplitude
  frequency: positiveFinite.max(1000).optional(), // Max 1000 Hz
  phase: normalized01.optional(), // 0-1 phase offset
  axis: z.enum(["x", "y", "z", "scale", "rotation"]).optional(),
  suggestedPath: boundedArray(ControlPointSchema, 10000).optional(), // Max 10k points
}).strict();

export type LayerMotionIntent = z.infer<typeof LayerMotionIntentSchema>;

// ============================================================================
// Motion Intent Result Schema (Main Schema)
// ============================================================================

/**
 * Motion Intent Result
 * Full AI response structure for motion suggestions
 */
export const MotionIntentResultSchema = z.object({
  description: z.string().min(1).max(MAX_DESCRIPTION_LENGTH).trim(),
  confidence: normalized01, // 0-1 confidence
  cameraIntents: boundedArray(CameraMotionIntentSchema, 100).optional(), // Max 100 camera intents
  layerIntents: boundedArray(LayerMotionIntentSchema, 1000).optional(), // Max 1k layer intents
  particleIntents: boundedArray(ParticleMotionIntentSchema, 100).optional(), // Max 100 particle intents
  splineIntents: boundedArray(SplineMotionIntentSchema, 100).optional(), // Max 100 spline intents
  rawResponse: z.string().max(100000).trim().optional(), // Max 100k chars for debugging
}).strict();

export type MotionIntentResult = z.infer<typeof MotionIntentResultSchema>;

// ============================================================================
// Motion Suggestion Schema (Alias for AI Response)
// ============================================================================

/**
 * Motion Suggestion Schema
 * Alias for MotionIntentResultSchema - used for AI response validation
 */
export const MotionSuggestionSchema = MotionIntentResultSchema;

export type MotionSuggestion = z.infer<typeof MotionSuggestionSchema>;

// ============================================================================
// Validation Helpers
// ============================================================================

export function validateMotionIntentResult(data: unknown): MotionIntentResult {
  return MotionIntentResultSchema.parse(data);
}

export function safeValidateMotionIntentResult(data: unknown) {
  return MotionIntentResultSchema.safeParse(data);
}

export function validateMotionSuggestion(data: unknown): MotionSuggestion {
  return MotionSuggestionSchema.parse(data);
}

export function safeValidateMotionSuggestion(data: unknown) {
  return MotionSuggestionSchema.safeParse(data);
}
